Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +634,463 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+    return false;
+}
+
+void FastStringifier::append(JSValue value)
+{
+    if (value.isNull()) {
+        if (remainingCapacity() < 4) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = 'n';
+        m_buffer[m_length + 1] = 'u';
+        m_buffer[m_length + 2] = 'l';
+        m_buffer[m_length + 3] = 'l';
+        m_length += 4;
+        return;
+    }
+
+    if (value.isTrue()) {
+        if (remainingCapacity() < 4) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = 't';
+        m_buffer[m_length + 1] = 'r';
+        m_buffer[m_length + 2] = 'u';
+        m_buffer[m_length + 3] = 'e';
+        m_length += 4;
+        return;
+    }
+
+    if (value.isFalse()) {
+        if (remainingCapacity() < 5) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = 'f';
+        m_buffer[m_length + 1] = 'a';
+        m_buffer[m_length + 2] = 'l';
+        m_buffer[m_length + 3] = 's';
+        m_buffer[m_length + 4] = 'e';
+        m_length += 5;
+        return;
+    }
+
+    if (value.isInt32()) {
+        auto number = value.asInt32();
+        auto length = lengthOfIntegerAsString(number);
+        if (remainingCapacity() < length) {
+            recordBufferFull();
+            return;
+        }
+        writeIntegerToBuffer(number, &m_buffer[m_length]);
+        m_length += length;
+        return;
+    }
+
+    if (!value.isCell()) {
+        recordFailure("value type"_s);
+        return;
+    }
+    auto& cell = *value.asCell();
+
+    switch (cell.type()) {
+    case StringType: {
+        auto& string = asString(&cell)->tryGetValue();
+        if (string.isNull()) {
+            recordFailure("String::tryGetValue"_s);
+            return;
+        }
+        if (!string.is8Bit()) {
+            recordFailure("16-bit string"_s);
+            return;
+        }
+        auto stringLength = string.length();
+        if (remainingCapacity() < 1 + stringLength + 1) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = '"';
+        auto* characters = string.characters8();
+        for (unsigned i = 0; i < stringLength; ++i) {
+            auto character = characters[i];
+            if (WTF::escapedFormsForJSON[character]) {
+                recordFailure("string character needs escaping"_s);
+                return;
+            }
+            m_buffer[m_length + 1 + i] = character;
+        }
+        m_buffer[m_length + 1 + stringLength] = '"';
+        m_length += 1 + stringLength + 1;
+        return;
+    }
+
+    case ObjectType:
+    case FinalObjectType: {
+        auto& object = *asObject(&cell);
+        if (object.isCallable()) {
+            recordFailure("callable object"_s);
+            return;
+        }
+        auto* ancestorClass = object.classInfo();
```

Comment:
Fetching `classInfo()` entails first fetching `structure()` in its implementation.  Consider pre-fetching `structure()` explicitly here (so that you can also use it after the loop below), and then fetch the classInfo  using `structure->classInfoForCells()`.  The computation of structure involves multiple instructions (though not slow), and it looks like you should have enough registers for the compiler to keep structure in a register.

Reply:
Done. Slightly sad that doing this causes me to lose the assertions from `JSCell::classInfo()`, but not so sad that I will add them.